<a href="https://colab.research.google.com/github/nanaaries313/Portfolio/blob/main/SVM_R.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Week 12: Support Vector Machines**

In [3]:
pkgs <- c("e1071","mlbench")
to_install <- pkgs[!pkgs %in% rownames(installed.packages())]
if (length(to_install) > 0) install.packages(to_install)

Installing packages into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependency ‘proxy’




In [4]:
library(e1071) # For the svm() function
library(mlbench) # For the glass data

Our familiar glass data can be obtained directly from the mlbench package

In [5]:
data(Glass, package="mlbench")
str(Glass)

'data.frame':	214 obs. of  10 variables:
 $ RI  : num  1.52 1.52 1.52 1.52 1.52 ...
 $ Na  : num  13.6 13.9 13.5 13.2 13.3 ...
 $ Mg  : num  4.49 3.6 3.55 3.69 3.62 3.61 3.6 3.61 3.58 3.6 ...
 $ Al  : num  1.1 1.36 1.54 1.29 1.24 1.62 1.14 1.05 1.37 1.36 ...
 $ Si  : num  71.8 72.7 73 72.6 73.1 ...
 $ K   : num  0.06 0.48 0.39 0.57 0.55 0.64 0.58 0.57 0.56 0.57 ...
 $ Ca  : num  8.75 7.83 7.78 8.22 8.07 8.07 8.17 8.24 8.3 8.4 ...
 $ Ba  : num  0 0 0 0 0 0 0 0 0 0 ...
 $ Fe  : num  0 0 0 0 0 0.26 0 0 0 0.11 ...
 $ Type: Factor w/ 6 levels "1","2","3","5",..: 1 1 1 1 1 1 1 1 1 1 ...


Glass data is more complicated data, so a method like SVM could be helpful

Split into test-train (resampling may be more appropriate for such small data, but that's not our focus today)

In [6]:
n=nrow(Glass)
set.seed(1234)
test = sample(1:n, as.integer(n/3))
test.data = Glass[test,]
train.data = Glass[-test,]

Run the svm() function with default parameters and construct the confusion matrix

In [8]:
svm.model = svm(Type ~ ., data = train.data)
CM = table(predict(svm.model,test.data),
           test.data$Type)
CM
sum(diag(CM))/sum (CM)
svm.model # What kernel is used by default? Options?

   
     1  2  3  5  6  7
  1 17 13  3  0  1  0
  2  4 16  2  2  1  0
  3  0  0  0  0  0  0
  5  0  2  0  3  1  0
  6  0  0  0  0  1  0
  7  0  0  0  0  0  5

[1] 0.5915493


Call:
svm(formula = Type ~ ., data = train.data)


Parameters:
   SVM-Type:  C-classification 
 SVM-Kernel:  radial 
       cost:  1 

Number of Support Vectors:  120


By default, SVM model uses the radial kernel. Cost is 1 (error or penalty, in case the data points are not in the right side of the hyperplane)

These are the two hyperparameters that we can change

In [9]:
#Total data points
n
#In the training data set
n*2/3

[1] 214

[1] 142.6667

120 is very high suport vectors numbers, so SVM must match it with very high dimensional space to get this numbers

In [10]:
?svm

Try a linear kernel

In [13]:
svm.model.linear = svm(Type ~ ., data = train.data,
                       kernel="linear")
CM.linear = table(predict(svm.model.linear,test.data),
           test.data$Type)
CM.linear
sum(diag(CM.linear))/sum (CM.linear)

   
     1  2  3  5  6  7
  1 17 14  4  0  0  0
  2  3 16  1  3  2  0
  3  1  0  0  0  0  0
  5  0  1  0  1  1  0
  6  0  0  0  0  1  0
  7  0  0  0  1  0  5

[1] 0.5633803

Try a polynomial kernel

In [15]:
svm.model.poly = svm(Type ~ ., data = train.data,
                       kernel="polynomial")
CM.poly = table(predict(svm.model.poly,test.data),
                  test.data$Type)
CM.poly
sum(diag(CM.poly))/sum (CM.poly)

svm.model.poly # What is the degree of the polynomial?

   
     1  2  3  5  6  7
  1 21 25  4  3  3  0
  2  0  4  1  1  1  0
  3  0  0  0  0  0  0
  5  0  2  0  1  0  0
  6  0  0  0  0  0  0
  7  0  0  0  0  0  5

[1] 0.4366197


Call:
svm(formula = Type ~ ., data = train.data, kernel = "polynomial")


Parameters:
   SVM-Type:  C-classification 
 SVM-Kernel:  polynomial 
       cost:  1 
     degree:  3 
     coef.0:  0 

Number of Support Vectors:  126


Try a simpler polynomial kernel

In [16]:
svm.model.poly.2 = svm(Type ~ ., data = train.data,
                       kernel="polynomial",
                       degree = 2)
CM.poly.2 = table(predict(svm.model.poly.2,test.data),
                  test.data$Type)
CM.poly.2

   
     1  2  3  5  6  7
  1 20 24  4  3  3  0
  2  1  7  1  1  1  0
  3  0  0  0  0  0  0
  5  0  0  0  1  0  0
  6  0  0  0  0  0  0
  7  0  0  0  0  0  5

 Try a more complex polynomial kernel

In [17]:
svm.model.poly.4 = svm(Type ~ ., data = train.data,
                       kernel="polynomial",
                       degree = 4)
CM.poly.4 = table(predict(svm.model.poly.4,test.data),
                  test.data$Type)
CM.poly.4

   
     1  2  3  5  6  7
  1 21 26  4  3  3  0
  2  0  3  0  1  1  0
  3  0  0  1  0  0  0
  5  0  2  0  1  0  0
  6  0  0  0  0  0  0
  7  0  0  0  0  0  5

Different kernel and parameters work well for different issues. We do want to experiement to find out the best model

 This is very interesting!

We've tried the default (radial basis), linear and polynomiall kernels. There is one left to try

In [18]:
svm.model.sigmoid = svm(Type ~ ., data = train.data,
                       kernel="sigmoid")
CM.sigmoid = table(predict(svm.model.sigmoid,test.data),
                   test.data$Type)
CM.sigmoid

   
     1  2  3  5  6  7
  1 19 19  4  0  1  0
  2  2 11  1  4  3  0
  3  0  0  0  0  0  0
  5  0  0  0  1  0  0
  6  0  1  0  0  0  0
  7  0  0  0  0  0  5

 What parameters can we try to tune?

In [19]:
svm.model.sigmoid$coef0 # Added constant in the kernel
svm.model.sigmoid$gamma # Multiplier in the kernel
svm.model.sigmoid$cost  # Penalty for violating constraints

svm.model.sigmoid = svm(Type ~ ., data = train.data,
                        kernel="sigmoid",
                        coef0 = 1,
                        gamma = 1)
CM.sigmoid = table(predict(svm.model.sigmoid,test.data),
                   test.data$Type)
CM.sigmoid

[1] 0

[1] 0.1111111

[1] 1

   
     1  2  3  5  6  7
  1 11 13  1  0  2  0
  2 10 15  4  2  1  1
  3  0  0  0  0  0  0
  5  0  0  0  1  0  0
  6  0  0  0  0  0  0
  7  0  3  0  2  1  4

Tuning has a large effect! You now have a huge amount of parameter tuning that you can do to find the best SVM

This is the first time that you are working with a method that will allow a great deal of tuning.

               Degree   Gamma  coef0   Cost
    Linear                              X
    Polynomial   X       X      X       X
    Radial Basis         X      X       X
    Sigmoid              X      X       X

My rule-of-thumb is to start by increasing and decreasing the default values, e.g., Degree = {2,3,4}, Cost = {0.1,1,10}.

This will already be a lot of options!

                Degree   Gamma  coef0   Cost
    Linear                               X      3
    Polynomial    X       X      X       X     81
    Radial Basis          X      X       X     27
    Sigmoid               X      X       X     27
    TOTAL                                     138
It is then a good idea to follow-up on the most promising
directions (e.g., try to tune the best kernel further).

You can use caret but make sure that it allows you to tune
everything that you want to tune!

There is also a tune.svm() function (in e1071). It works
similarly to the caret train() function.

##**Why do we tune?**
For better model

Find the kernel that works the best. And then tune for that particular kernel.

Closer to the training data? low bias. high variance.

Better model = bias-variance trade-off

How do I know if I got this right?

In [20]:
svm.model = svm(Type ~ ., data = train.data)
CM = table(predict(svm.model,test.data),
           test.data$Type)
CM
sum(diag(CM))/sum (CM)

#what if we look at the training error?
CM.train = table(predict(svm.model,train.data),
           train.data$Type)
CM.train
sum(diag(CM.train))/sum (CM.train)

   
     1  2  3  5  6  7
  1 17 13  3  0  1  0
  2  4 16  2  2  1  0
  3  0  0  0  0  0  0
  5  0  2  0  3  1  0
  6  0  0  0  0  1  0
  7  0  0  0  0  0  5

[1] 0.5915493

   
     1  2  3  5  6  7
  1 47 10  9  0  1  1
  2  2 35  3  0  0  0
  3  0  0  0  0  0  0
  5  0  0  0  8  0  0
  6  0  0  0  0  4  0
  7  0  0  0  0  0 23

[1] 0.8181818

Boosted tree or deep learning are go-to methods for difficult datasets

In [21]:
svm.model = svm(Type ~ ., data = train.data,
                cost = 0.01)
CM = table(predict(svm.model,test.data),
           test.data$Type)
CM
sum(diag(CM))/sum (CM)

#what if we look at the training error?
CM.train = table(predict(svm.model,train.data),
           train.data$Type)
CM.train
sum(diag(CM.train))/sum (CM.train)

   
     1  2  3  5  6  7
  1 21 31  5  5  4  5
  2  0  0  0  0  0  0
  3  0  0  0  0  0  0
  5  0  0  0  0  0  0
  6  0  0  0  0  0  0
  7  0  0  0  0  0  0

[1] 0.2957746

   
     1  2  3  5  6  7
  1 49 45 12  8  5 24
  2  0  0  0  0  0  0
  3  0  0  0  0  0  0
  5  0  0  0  0  0  0
  6  0  0  0  0  0  0
  7  0  0  0  0  0  0

[1] 0.3426573

This is generalizing as much as we can. This is exactly the same thing as KNN with high enough K, we predict the majority class

In [22]:
svm.model = svm(Type ~ ., data = train.data,
                cost = 100)
CM = table(predict(svm.model,test.data),
           test.data$Type)
CM
sum(diag(CM))/sum (CM)

#what if we look at the training error?
CM.train = table(predict(svm.model,train.data),
           train.data$Type)
CM.train
sum(diag(CM.train))/sum (CM.train)

   
     1  2  3  5  6  7
  1 16 13  2  0  0  0
  2  4 15  1  2  1  0
  3  1  1  2  0  0  0
  5  0  2  0  3  1  0
  6  0  0  0  0  2  0
  7  0  0  0  0  0  5

[1] 0.6056338

   
     1  2  3  5  6  7
  1 46  2  0  0  0  0
  2  2 42  0  0  0  0
  3  1  1 12  0  0  0
  5  0  0  0  8  0  0
  6  0  0  0  0  5  0
  7  0  0  0  0  0 24

[1] 0.958042

In [24]:
svm.model = svm(Type ~ ., data = train.data,
                cost = 0.5)
CM = table(predict(svm.model,test.data),
           test.data$Type)
CM
sum(diag(CM))/sum (CM)

#what if we look at the training error?
CM.train = table(predict(svm.model,train.data),
           train.data$Type)
CM.train
sum(diag(CM.train))/sum (CM.train)

   
     1  2  3  5  6  7
  1 17 16  3  1  1  0
  2  4 15  2  3  3  0
  3  0  0  0  0  0  0
  5  0  0  0  1  0  0
  6  0  0  0  0  0  0
  7  0  0  0  0  0  5

[1] 0.5352113

   
     1  2  3  5  6  7
  1 48 10 10  0  1  1
  2  1 35  2  4  4  1
  3  0  0  0  0  0  0
  5  0  0  0  4  0  0
  6  0  0  0  0  0  0
  7  0  0  0  0  0 22

[1] 0.7622378

In [36]:
svm.model = svm(Type ~ ., data = train.data,
                cost = 1,
                kernel = "linear")
CM = table(predict(svm.model,test.data),
           test.data$Type)
CM
sum(diag(CM))/sum (CM)

#what if we look at the training error?
CM.train = table(predict(svm.model,train.data),
           train.data$Type)
CM.train
sum(diag(CM.train))/sum (CM.train)

   
     1  2  3  5  6  7
  1 17 14  4  0  0  0
  2  3 16  1  3  2  0
  3  1  0  0  0  0  0
  5  0  1  0  1  1  0
  6  0  0  0  0  1  0
  7  0  0  0  1  0  5

[1] 0.5633803

   
     1  2  3  5  6  7
  1 40 14  6  0  0  1
  2  9 31  6  3  0  0
  3  0  0  0  0  0  0
  5  0  0  0  5  0  0
  6  0  0  0  0  5  0
  7  0  0  0  0  0 23

[1] 0.7272727

In [34]:
svm.model = svm(Type ~ ., data = train.data,
                cost = 100,
                kernel = "linear")
CM = table(predict(svm.model,test.data),
           test.data$Type)
CM
sum(diag(CM))/sum (CM)

#what if we look at the training error?
CM.train = table(predict(svm.model,train.data),
           train.data$Type)
CM.train
sum(diag(CM.train))/sum (CM.train)

   
     1  2  3  5  6  7
  1 15 11  2  0  0  0
  2  4 15  1  3  2  0
  3  2  5  2  0  0  0
  5  0  0  0  2  1  0
  6  0  0  0  0  1  0
  7  0  0  0  0  0  5

[1] 0.5633803

   
     1  2  3  5  6  7
  1 34  8  0  0  0  0
  2 13 35  2  0  0  0
  3  2  2 10  0  0  0
  5  0  0  0  8  0  0
  6  0  0  0  0  5  0
  7  0  0  0  0  0 24

[1] 0.8111888